### ISSUE TITLE
Minimize security exposures via bot, clarify CLI usage

### ISSUE THREAD

**Issue by nealmcb** [all-contributors/all-contributors#284]
The current README doesn't clarify usage other than by pointing at  the @all-contributors bot ([Bot Overview · All Contributors](https://allcontributors.org/docs/en/bot/overview)).

That bot documentation doesn't note the manifold security implications of adding the bot as a committer to a user's repository. E.g. unless I'm misunderstanding something, by granting "Write access to code" to the bot, if the bot is hacked, all the repos it can update can be hacked. That sounds like a serious issue, and makes you a big target, right?

So I think the README should be updated to point out the CLI, and recommend it, especially for low-churn repos.

And the bot documentation should be changed to warn of the implications.

In fact, couldn't the bot be changed to just create PRs, which must be merged by hand? That would put the security issues back in the hands of the proper maintainers.

See an example of some other security measures you could take and assurances you could provide to users at [renovate/data-security.md at master · renovatebot/renovate](https://github.com/renovatebot/renovate/blob/master/website/docs/data-security.md). Note that they seem talk about some useful protections, but I don't know if they are actually doing things securely, and that depends on lots of things. Security is hard!

GitHub could of course provide more fine-grained permissions so you could be more secure. See e.g. [GitHub's Permission System is Flawed · Issue #113 · dear-github/dear-github](https://github.com/dear-github/dear-github/issues/113). Beyond that, for this project, it would be great if the permissions could limit write access to a single directory.

Thanks!

**Comment by Berkmann18** [comment]
Thanks for your issue.
The bot is preferred as it automates the process more (although it lacks a few things that are implemented or nearly implemented on the CLI).

> In fact, couldn't the bot be changed to just create PRs, which must be merged by hand? That would put the security issues back in the hands of the proper maintainers.

That's actually what the bot does.

Feel free to submit a PR tho.

> Beyond that, for this project, it would be great if the permissions could limit write access to a single directory.

I'm not sure that's possible as of yet as I've never seen permissions based on specific directories within a repo (I might be wrong).

**Comment by nealmcb** [comment]
Thanks for the quick response.

I'm glad to hear that by design the bot doesn't merge the changes.

But unless I'm mistaken, the security issue remains that given the permissions that users are required to give it, the bot *could* insert malicious content anywhere in the repo, without any approval by others. And it could do that for all of the 2000+ repos it has access to, and I wonder if that it could also release, or trigger CI systems to automatically release, new poisoned versions. I think it's important to document that threat, and the steps that you and your users can take to mitigate it.  E.g. audits of what files have been changed in commits authored by the bot, audits of any unusual actions triggered by the bot, etc.

As a more effective remedy, I meant to clarify that I was thinking that the bot could instead pull a fork of the repo into its own organization, and mount the PR from there, eliminating any need for write permission on the repo.

What would it take to do that?

**Comment by Berkmann18** [comment]
> But unless I'm mistaken, the security issue remains that given the permissions that users are required to give it, the bot could insert malicious content anywhere in the repo, without any approval by others. And it could do that for all of the 2000+ repos it has access to, and I wonder if that it could also release, or trigger CI systems to automatically release, new poisoned applications and libraries.

Given how the bot works, it would have to be a huge attack that someone would need to pass through and no one can directly push to master or without having a PR reviewed by a human.

> I think it's important to document that threat, and the steps that you and your users can take to mitigate it. E.g. audits of what files have been changed in commits authored by the bot, audits of any unusual actions triggered by the bot, etc.

True, I was thinking of getting `snyk` set up but it's costly and the whole team do that voluntarily and we don't get any funds so investing in such tools is not an option (for now).
I personally use `npm/yarn audit` and `snyk` so I always try to fix vulnerabilities I find.

> As a more effective remedy, I meant to clarify that I was thinking that the bot could instead pull a fork of the repo into its own organization, and mount the PR from there, eliminating any need for write permission on the user's repo.

That seems to be a good idea although I'm not sure if having the bot managing forks as opposed to just its branches would be practical and maintainable.

@all-contributors/core What do you think?

**Comment by nealmcb** [comment]
Thanks again.
Indeed, I don't expect it would be easy to officially get a PR thru your team that makes the bot malicious.
I expect it would be easier to compromise the bot itself via accounts with permission to control it, the infrastructure it runs on, or other attack vectors, of which there are many, many of which have been compromised. The renovate link above covers mitigations for lots more risks.

**Comment by Berkmann18** [comment]
That renovate doc looks good and I agree it would be nice to have one for all-contributors/all-contributors-bot.

You are more than welcome to submit more ideas or start a PR for that.

**Comment by adamralph** [comment]
> > In fact, couldn't the bot be changed to just create PRs, which must be merged by hand? That would put the security issues back in the hands of the proper maintainers.
> 
> That's actually what the bot does.

Then why does it need write access?

**Comment by gr2m** [comment]
It cannot create pull requests without the [`contents` permission](https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-contents). GitHub App permissions are not as granular as we wish they would be, but as long as you have branch protections in place, which you should anyway if you are concerned about security, the app cannot update the protected branches directly, nor can it merge its own pull requests unless the pull request status is green.

**Comment by adamralph** [comment]
Is that because it pushes the pull request branch directly to the target repo? If so, could it not use a fork instead?

**Comment by gr2m** [comment]
> Is that because it pushes the pull request branch directly to the target repo?

It does not push a branch, it does not use `git`. It uses GitHub's REST API.

>  If so, could it not use a fork instead?

A GitHub App has no account it could fork to.

**Comment by adamralph** [comment]
> It does not push a branch, it does not use `git`. It uses GitHub's REST API.

Same end effect though: it requires write access to create a branch in the target repo.
 
> A GitHub App has no account it could fork to.

I realise I may be crossing into the realm of elaborate solutions, but could you not create an account specifically for that purpose?

**Comment by gr2m** [comment]
> I realise I may be crossing into the realm of elaborate solutions, but could you not create an account specifically for that purpose?

It's not that easy. For example, that wouldn't work for private repositories. You'd need to add the user to the repository and then you had an even bigger security problem.

This is really something that GitHub needs to fix by making the permissions more granular, we cannot work around it with a project we work on in our after hours :)

Another approach that would take better advantage of GitHub's permissions is to restrict which files the app can updated by the app, but files are currently configurable, and the README files can have different file names, too.

So anyway, I don't really see a problem. If you are concerned about security, you want to setup good branch protection rules. If you setup good branch protection rules, the app can't do anything harmful.

**Comment by adamralph** [comment]
> This is really something that GitHub needs to fix by making the permissions more granular

I agree, and I sympathise with any reluctance to work around it in the meantime.

Just for the sake of completeness:

> that wouldn't work for private repositories

I believe it would if the app had _read_ permissions on private repos, but again, I guess GitHub's lack of granular permissions may get in the way.

> If you setup good branch protection rules, the app can't do anything harmful.

True, but that means I would have to set up branch protection for _every_ branch in my repo. That's fine for "permanent" branches such `master`/`main`/`release-{x}` etc. but it's impractical for temporary branches, such as those used for PRs.
